storage.go

```diff
@@ -2,35 +2,222 @@ package gdrive
+		if i == 0 {
+			fileId, err = s.searchContentInDir(ctx, "root", pathUnits[i])
+		}
+		fileId, err = s.searchContentInDir(ctx, fileId, pathUnits[i])
+	}
+
+	if err != nil {
+		return "", err
+	}
+	return fileId, nil
+}
+
+func (s *Storage) rawListFiles(ctx context.Context, searchArg string) ([]*drive.File, error) {
+	var fs []*drive.File
+	pageToken := ""
+	for {
+		q := s.service.Files.List().Context(ctx).Q(searchArg).Fields("id", "name", "mimeType")
+		// If we have a pageToken set, apply it to the query
+		if pageToken != "" {
+			q = q.PageToken(pageToken)
+		}
+		r, err := q.Do()
+		if err != nil {
+			return fs, err
+		}
+		fs = append(fs, r.Files...)
+		pageToken = r.NextPageToken
+		if pageToken == "" {
+			break
+		}
+	}
+	return fs, nil
 }
 
 func (s *Storage) read(ctx context.Context, path string, w io.Writer, opt pairStorageRead) (n int64, err error) {
-	panic("not implemented")
+	fileId, err := s.pathToId(ctx, path)
+	if err != nil {
+		return 0, err
+	}
+	f, err := s.service.Files.Get(fileId).Context(ctx).Download()
+	if err != nil {
+		return 0, err
+	}
+
+	if opt.HasIoCallback {
+		iowrap.CallbackReadCloser(f.Body, opt.IoCallback)
+	}
+
+	return io.Copy(w, f.Body)
+}
+
+// We pass the fileId of the directory, and it returns the fileId of the content in the directory
+func (s *Storage) searchContentInDir(ctx context.Context, dirId string, contentName string) (fileId string, err error) {
+	fileList, err := s.listFilesInDir(ctx, dirId)
+
+	if err != nil {
+		return " ", err
+	}
+
+	for _, file := range fileList {
+		// What if the file does not exist???
+		if file.Name == contentName {
+			fileId = file.Id
+		}
+	}
+
+	if len(fileId) == 0 {
+		return "", err
+	}
+	return fileId, nil
 }
 
 func (s *Storage) stat(ctx context.Context, path string, opt pairStorageStat) (o *Object, err error) {
-	panic("not implemented")
+	_, err = s.pathToId(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+	rp := s.getAbsPath(path)
+	o = s.newObject(true)
+	o.ID = rp
+	o.Path = path
+	return o, nil
 }
 
+// First we need make sure this file is not exist
+// If it is, then we upload it
+// If the path is like `foo.txt`, then we upload it directly.
+// But if the path is like `foo/bar.txt`, we need create the directory one by one
 func (s *Storage) write(ctx context.Context, path string, r io.Reader, size int64, opt pairStorageWrite) (n int64, err error) {
-	panic("not implemented")
+	r = io.LimitReader(r, size)
+
+	if opt.HasIoCallback {
+		r = iowrap.CallbackReader(r, opt.IoCallback)
+	}
+	_, err = s.pathToId(ctx, path)
+	var fileId string
+	if err != nil {
+		// upload
+		tmp := strings.Split(path, "/")
+		for i := 0; i < len(tmp); i++ {
+			if i == 0 {
+				fileId, err = s.createDir(ctx, tmp[0], "root")
+			} else if i != len(tmp)-1 {
+				fileId, err = s.createDir(ctx, tmp[i], fileId)
+			} else {
+				file := &drive.File{Name: tmp[i]}
+				_, err = s.service.Files.Create(file).Context(ctx).Media(r).Do()
+			}
```

Comment:
I guess I have misunderstood something. If user calls Write("a/b/c.txt"), but directory a/b is not exist, should we create them for user? If not, these lines could be simply deleted. Ping @Xuanwo to have a look

Reply:
>  If user calls `Write("a/b/c.txt")`, but directory `a/b` is not exist, should we create them for user?

Yes, we need to create dirs for users.

Reply:
OK, I will do a quick fix for this.